### PR TITLE
upload track to cloud

### PR DESCRIPTION
**Pull request by alisa911**

### CODE REVIEW

**Review by RZR-UA — APPROVED** [review]

**Comment by alisa911** [comment]
https://github.com/osmandapp/OsmAnd-Issues/issues/2204

File: map/src/manager/track/TracksManager.js [hunk 1]
```diff
@@ -788,49 +560,14 @@ export function findGroupByName(groups, groupName) {
...
-            newGpxFiles[file.name][`${t}`] = track[t];
-        });
-        newGpxFiles[file.name].analysis = TracksManager.prepareAnalysis(newGpxFiles[file.name].analysis);
-        ctx.setGpxFiles(newGpxFiles);
-        ctx.setSelectedGpxFile(Object.assign({}, newGpxFiles[file.name]));
-    }
+export function hasExistTrack(name, folder, tracks) {
+    const folderName = folder.title ? folder.title : folder;
+    const foundFolder = findGroupByName(tracks, folderName);
+    return foundFolder ? foundFolder.files.find((f) => TracksManager.prepareName(f.name) === name) : false;
```
Comment (RZR-UA) [inline map/src/manager/track/TracksManager.js:571]: 1) maybe better name would be folderHasTrack() or isTrackExists()
2) as this is boolean function, better to use "some()" not "find()"
3) please replace find->some in the parent (for local tracks)

File: map/src/menu/tracks/TrackGroupFolder.jsx [hunk 2]
```diff
@@ -27,7 +27,7 @@ export default function TrackGroupFolder({ folder }) {
 
     useEffect(() => {
         if (ctx.tracksGroups) {
-            let found = findGroupByName(ctx.tracksGroups, group.fullName);
+            let found = findGroupByName(ctx.tracksGroups, group.fullName, 'fullName');
```
Comment (RZR-UA) [inline map/src/menu/tracks/TrackGroupFolder.jsx:30]: What is the `'fullName'` in parameters?

**Review by RZR-UA — CHANGES_REQUESTED** [review]

File: map/src/manager/track/SaveTrackManager.js [hunk 3]
```diff
@@ -166,6 +155,43 @@ export function saveTrackToLocalStorage({ ctx, track }) {
...
+    };
+    //save empty file with new folder's name
+    const res = await apiPost(`${process.env.REACT_APP_USER_API_SITE}/mapapi/upload-file`, data, { params });
+    if (res && res?.data?.status === 'ok') {
+        refreshGlobalFiles(ctx, params).then();
+        return true;
+    }
+}
+
+async function refreshGlobalFiles(ctx, params) {
```
Comment (RZR-UA) [inline map/src/manager/track/SaveTrackManager.js:180]: Slightly unclear what are `params` and why only 1 of of params is used. Better to rename params to smth such as `currentFile` or `currentFileName` and additionally change caller's `params` to `params?.name`.

File: map/src/manager/track/SaveTrackManager.js [hunk 4]
```diff
@@ -166,6 +155,43 @@ export function saveTrackToLocalStorage({ ctx, track }) {
     });
 }
 
+export async function saveEmptyTrack(folderName, ctx) {
+    //create empty file
+    const convertedData = new TextEncoder().encode('');
+    const zippedResult = require('pako').gzip(convertedData, { to: 'Uint8Array' });
+    const convertedZipped = zippedResult.buffer;
+    const oMyBlob = new Blob([convertedZipped], { type: 'gpx' });
```
Comment (RZR-UA) [inline map/src/manager/track/SaveTrackManager.js:163]: Is it really required to use empty gzip file here? Is the upload-file API requires this?

File: map/src/menu/tracks/actions/TracksHeader.jsx [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+    const DialogAddFolder = () => {
+        const [folderNameError, setFolderNameError] = useState('');
+        const [folderName, setFolderName] = useState('');
+
+        function validationFolderName(name) {
+            if (!name || name === '' || name.trim().length === 0) {
+                setFolderNameError('Empty folder name!');
+            } else if (isFolderExist(name)) {
+                setFolderNameError('Folder with same name already exist!');
```
Comment (RZR-UA) [inline map/src/menu/tracks/actions/TracksHeader.jsx:46]: Folder already exists.

File: map/src/menu/tracks/actions/TracksHeader.jsx [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+        return trackGroup.subfolders.some((folder) => folder.name === name);
+    }
+
+    const DialogAddFolder = () => {
+        const [folderNameError, setFolderNameError] = useState('');
+        const [folderName, setFolderName] = useState('');
+
+        function validationFolderName(name) {
+            if (!name || name === '' || name.trim().length === 0) {
+                setFolderNameError('Empty folder name!');
```
Comment (RZR-UA) [inline map/src/menu/tracks/actions/TracksHeader.jsx:44]: Empty folder name.

File: map/src/menu/tracks/actions/TracksHeader.jsx [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+    }
+
+    function prevTrackMenu() {
+        ctx.openTrackGroups.pop();
+        ctx.setOpenTrackGroups([...ctx.openTrackGroups]);
+    }
+
+    function isFolderExist(name) {
+        return trackGroup.subfolders.some((folder) => folder.name === name);
+    }
```
Comment (RZR-UA) [inline map/src/menu/tracks/actions/TracksHeader.jsx:36]: Doesn't work in top-level folder.

File: map/src/menu/tracks/actions/TracksHeader.jsx [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+            const newFolderName = `${folderPart}${folderName}`;
+            saveEmptyTrack(newFolderName, ctx).then();
+            setOpenAddFolderDialog(false);
+        }
+
+        return (
+            <Dialog open={true} onClose={() => setOpenAddFolderDialog(false)}>
+                <DialogTitle className={dialogStyles.title}>Add new folder</DialogTitle>
+                <DialogContent className={dialogStyles.content}>
+                    <TextField
```
Comment (RZR-UA) [inline map/src/menu/tracks/actions/TracksHeader.jsx:68]: 1) please run `addFolder()` by Enter key with onKeyPress handler

2) please use `label` or remove unused empty space on the top of the input field

**Review by RZR-UA — APPROVED** [review]

**Review by RZR-UA — APPROVED** [review]

**Review by RZR-UA — APPROVED** [review]

**Pull request merged**